### PR TITLE
Added iterator store for db intrinsic secondary keys in RocksDB.

### PR DESCRIPTION
**Pull request by brianjohnson5972**
## Change Description
Added an iterator store for secondary key lookup of db instrinsics in a RocksDB backing store.  This has similar requirements to db_chainbase_iter_cache, except that for chainbase the table and secondary key info was stored in an external object in a multi-index container, while this storage is all internal to the cache. Also, changed db_chainbase_iter_cache to db_chainbase_iter_store to be more accurate.

## Change Type
**Select ONE**
- [ ] Documentation

- [ ] Stability bug fix

- [x] Other

- [ ] Other - special case

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+template<typename T>
+struct secondary_key {
+   int       table_ei;
+   T         secondary;
+   uint64_t  primary;
+   uint64_t  payer;
+};
+
+struct unique_table_hash {
```
Comment (TimothyBanks) [inline libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp:22]: I need my memory refreshed on this, but couldn't we add a partial specialization of `std::hash`?  I thought the standard allowed for partial specialization of class templates as long as it depended on a user defined type.

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+         return _iterator_to_object.size() - 1;
+      }
+
+   private:
+      void validate_object_iterator(int iterator, const char* explanation_for_no_end_iterators) const {
+         EOS_ASSERT( iterator != _invalid_iterator, invalid_table_iterator, "invalid iterator" );
+         EOS_ASSERT( iterator >= 0, table_operation_not_permitted, explanation_for_no_end_iterators );
+         EOS_ASSERT( (size_t)iterator < _iterator_to_object.size(), invalid_table_iterator, "iterator out of range" );
+      }
+      unordered_map<unique_table, int, unique_table_hash > _table_cache;
```
Comment (swatanabe-b1) [inline libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp:144]: `unordered_map` is vulnerable to DOS attacks caused by hash collisions.

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+struct secondary_key {
+   int       table_ei;
+   T         secondary;
+   uint64_t  primary;
+   uint64_t  payer;
+};
+
+struct unique_table_hash {
+   std::size_t operator()(const unique_table& t) const {
+      return std::hash<name>()(t.contract) ^ std::hash<name>()(t.scope) ^ std::hash<name>()(t.table);
```
Comment (swatanabe-b1) [inline libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp:24]: `contract == scope` is fairly common and will cause them to cancel.

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp (hunk 1 above)
Comment (brianjohnson5972) [reply libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp:22]: dropping hash and switching to map, so this is a moot point

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp (hunk 2 above)
Comment (brianjohnson5972) [reply libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp:144]: changing to map

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+   return lhs.table_ei == rhs.table_ei &&
+          lhs.secondary == rhs.secondary &&
+          lhs.primary == rhs.primary;
+   // ignoring payer for uniqueness
+}
+
+template<typename SecondaryKey>
+class db_key_value_iter_cache {
+   public:
+      const int _invalid_iterator = -1;
```
[the remark this answers is not in the document]
Comment (brianjohnson5972) [reply libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp:52]: In the interest of how the code looks, I went with a constexpr method, then we get compile time, but not having to use it in reference to the long name plus template type.  I'm glad to change it all the way if that is still preferred.

File: libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp (hunk 3 above)
Comment (brianjohnson5972) [reply libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp]: Please verify you are ok with secondary_key operator<

Comment (swatanabe-b1) [reply libraries/chain/include/eosio/chain/backing_store/db_key_value_iter_cache.hpp]: operator< can be implemented using std::tie, which is easier to understand.

**Pull request merged**